examples/server/server.cpp

```diff
@@ -2178,6 +2208,249 @@ static void server_params_parse(int argc, char **argv, server_params &sparams,
     }
 }
 
+
+static std::string random_string() {
+  std::string str(
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
+
+  std::random_device rd;
+  std::mt19937 generator(rd());
+
+  std::shuffle(str.begin(), str.end(), generator);
+
+  return str.substr(0, 32); // assumes 32 < number of characters in str
+}
+
+static std::string gen_chatcmplid() {
+  std::stringstream chatcmplid;
+  chatcmplid << "chatcmpl-" << random_string();
+  return chatcmplid.str();
+}
+
+std::string format_chatml(std::vector<json> messages) {
+
+  std::ostringstream chatml_msgs;
+
+  // iterate the array
+  for (auto it = messages.begin(); it != messages.end(); ++it) {
+    chatml_msgs << "<|im_start|>"
+                << json_value(*it, "role", std::string("user")) << '\n';
+    chatml_msgs << json_value(*it, "content", std::string(""))
+                << "<|im_end|>\n";
+  }
+
+  chatml_msgs << "<|im_start|>assistant" << '\n';
+
+  return chatml_msgs.str();
+}
+
```

Comment:
I wonder if we should enable tokenization of special tokens by default in `sever` example:

https://github.com/ggerganov/llama.cpp/blob/8e672efe632bb6a7333964a255c4b96f018b9a65/examples/server/server.cpp#L610-L653

Here, the calls to `llama_tokenize` currently default to `special == false` because there are some edge cases where the user could be using html tags such as `<s>` and they would be tokenized incorrectly as special tokens.

Technically, the correct solution is the one we have in `main`:

https://github.com/ggerganov/llama.cpp/blob/8e672efe632bb6a7333964a255c4b96f018b9a65/examples/main/main.cpp#L787-L792

But this seems a bit tricky to implement in `server`.

I think there is no point in sacrificing ChatML compat just for that extreme edge case, so maybe we should use `special == true` in `server`.

@staviq and others, any thoughts?

Reply:
Can we enable it only in the context of /v1/chat/completions (and maybe other future) OpenAI compatible APIs?
Can we give it a whitelist of allowed special tokens? ChatML tokens are special due to "|"-character and won't clash with HTML tags.

Preferably, I would leave this decision & implementation to core contributors.

Reply:
There are many options and we can do anything we like, though it's a bit difficult for me to make the decision. There are so many variations of these templates and tokenization is incredibly over-complicated in general.

> Can we enable it only in the context of /v1/chat/completions (and maybe other future) OpenAI compatible APIs?

Yes, let's do that.

> Can we update tokenize to be able to give it a whitelist of allowed special tokens?

Not sure if this is needed. What issue would it fix?
The model already knows internally which are the special tokens.